x/gamm/twap/store.go

```diff
@@ -15,17 +24,121 @@ func (k Keeper) trackChangedPool(ctx sdk.Context, poolId uint64) {
+func (k Keeper) deleteHistoricalRecord(ctx sdk.Context, twap types.TwapRecord) {
+	store := ctx.KVStore(k.storeKey)
+	key1 := types.FormatHistoricalTimeIndexTWAPKey(twap.Time, twap.PoolId, twap.Asset0Denom, twap.Asset1Denom)
+	key2 := types.FormatHistoricalPoolIndexTWAPKey(twap.PoolId, twap.Time, twap.Asset0Denom, twap.Asset1Denom)
+	store.Delete(key1)
+	store.Delete(key2)
+}
+
+// getMostRecentRecordStoreRepresentation returns the most recent twap record in the store
+// for the provided (pool, asset0, asset1) triplet.
+// Its called store representation, because most recent record can refer to it being
+// interpolated to the current block time, or after events in this block.
+// Neither of which apply to the record this returns.
+func (k Keeper) getMostRecentRecordStoreRepresentation(ctx sdk.Context, poolId uint64, asset0Denom string, asset1Denom string) (types.TwapRecord, error) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.FormatMostRecentTWAPKey(poolId, asset0Denom, asset1Denom)
+	bz := store.Get(key)
+	return types.ParseTwapFromBz(bz)
+}
+
+// getMostRecentRecordStoreRepresentation returns all most recent twap records
+// (in state representation) for the provided pool id.
+func (k Keeper) getAllMostRecentRecordsForPool(ctx sdk.Context, poolId uint64) ([]types.TwapRecord, error) {
+	store := ctx.KVStore(k.storeKey)
+	return types.GetAllMostRecentTwapsForPool(store, poolId)
+}
+
+// storeNewRecord stores a record, in both the most recent record store and historical stores.
+func (k Keeper) storeNewRecord(ctx sdk.Context, twap types.TwapRecord) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.FormatMostRecentTWAPKey(twap.PoolId, twap.Asset0Denom, twap.Asset1Denom)
+	osmoutils.MustSet(store, key, &twap)
+	k.storeHistoricalTWAP(ctx, twap)
+}
+
+// getRecordAtOrBeforeTime on a given input (id, t, asset0, asset1)
+// returns the TWAP record from state for (id, t', asset0, asset1),
+// where t' is such that:
+// * t' <= t
+// * there exists no t'' in state, where t'' < t'
+//
+// This returns an error if:
+// * there is no historical record in state at or before t
+//   - Occurs if t is older than pruning period, or pool creation date.
+// * there is no record for the asset pair (asset0, asset1) in particular
+//   - e.g. asset not in pool, or provided in wrong order.
+func (k Keeper) getRecordAtOrBeforeTime(ctx sdk.Context, poolId uint64, t time.Time, asset0Denom string, asset1Denom string) (types.TwapRecord, error) {
+	store := ctx.KVStore(k.storeKey)
+	// We make an iteration from time=t + 1ns, to time=0 for this pool.
+	// Note that we cannot get any time entries from t + 1ns, as the key would be `prefix|t+1ns`
+	// and the end for a reverse iterator is exclusive. Thus the largest key that can be returned
+	// begins with a prefix of `prefix|t`
+	startKey := types.FormatHistoricalPoolIndexTimePrefix(poolId, time.Unix(0, 0))
+	endKey := types.FormatHistoricalPoolIndexTimePrefix(poolId, t.Add(time.Nanosecond))
```

Comment:
Example from one of the test cases:
```
"historical_pool_index|1|2009-11-10T23:00:00.000000001|"
```

Is key separator needed at the end of the key if time format is standardized? Seems like a redundant byte in db

Reply:
This is the prefix though? We iterate based off entries with the prefix, we need that separator to get all assets.

We could claim that its not needed due to fixed size times, but I'd rather not rely on that too explicitly for simplicity.

A bigger saving would be reconsidering using a fixed size unix time serialization than the key separator.

Reply:
https://github.com/osmosis-labs/osmosis/issues/2251